lib/OptimizationODE/src/OptimizationODE.jl

```diff
@@ -2,58 +2,189 @@ module OptimizationODE
+        temp = similar(u)
+        f_mass!(temp, u, p, 0.0)
+        res .= temp
+    end
+    u0_ext = vcat(u0, zeros(m))
+    prob_nl = NonlinearProblem(resid!, u0_ext, p)
+    sol_nl = solve(prob_nl, Newton(); tol = 1e-8, maxiters = 100000,
+        callback = cache.callback, progress = get(cache.solver_args, :progress, false))
+    u_ext = sol_nl.u
+    return u_ext[1:n], sol_nl.retcode
+end
+
+
+function get_solver_type(opt::DAEOptimizer)
+    if opt.solver isa Union{Rodas5, RadauIIA5, ImplicitEuler, Trapezoid}
+        return :mass_matrix
+    else
+        return :indexing
+    end
+end
 
-    dt    = cache.opt.dt
-    maxit = get(cache.solver_args, :maxiters, 1000)
+function handle_parameters(p)
+    if p isa SciMLBase.NullParameters
+        return Float64[]
+    else
+        return p
+    end
+end
+
+function setup_progress_callback(cache, solve_kwargs)
+    if get(cache.solver_args, :progress, false)
+        condition = (u, t, integrator) -> true
+        affect! = (integrator) -> begin
+            u_opt = integrator.u isa AbstractArray ? integrator.u : integrator.u.u
+            cache.solver_args[:callback](u_opt, integrator.p, integrator.t)
+        end
+        cb = DiscreteCallback(condition, affect!)
+        solve_kwargs[:callback] = cb
+    end
+    return solve_kwargs
+end
 
+function finite_difference_jacobian(f, x; ϵ = 1e-8)
+    n = length(x)
+    fx = f(x)
+    if fx === nothing
+        return zeros(eltype(x), 0, n)
+    elseif isa(fx, Number)
+        J = zeros(eltype(fx), 1, n)
+        for j in 1:n
+            xj = copy(x)
+            xj[j] += ϵ
+            diff = f(xj)
+            if diff === nothing
+                diffval = zero(eltype(fx))
+            else
+                diffval = diff - fx
+            end
+            J[1, j] = diffval / ϵ
+        end
+        return J
+    else
+        m = length(fx)
+        J = zeros(eltype(fx), m, n)
+        for j in 1:n
+            xj = copy(x)
+            xj[j] += ϵ
+            fxj = f(xj)
+            if fxj === nothing
+                @inbounds for i in 1:m
+                    J[i, j] = -fx[i] / ϵ
+                end
+            else
+                @inbounds for i in 1:m
+                    J[i, j] = (fxj[i] - fx[i]) / ϵ
+                end
+            end
+        end
+        return J
+    end
+end
+
+function SciMLBase.__solve(
+    cache::OptimizationCache{F,RC,LB,UB,LC,UC,S,O,D,P,C}
+    ) where {F,RC,LB,UB,LC,UC,S,O<:Union{ODEOptimizer,DAEOptimizer},D,P,C}
+
+    dt = get(cache.solver_args, :dt, nothing)
+    maxit = get(cache.solver_args, :maxiters, nothing)
+    differential_vars = get(cache.solver_args, :differential_vars, nothing)
     u0 = copy(cache.u0)
-    p  = cache.p
+    p = handle_parameters(cache.p)  # Properly handle NullParameters
 
+    if cache.opt isa ODEOptimizer
+        return solve_ode(cache, dt, maxit, u0, p)
+    else
+        solver_method = get_solver_type(cache.opt)
+        if solver_method == :mass_matrix
```

Comment:
The actual check is `alg isa DAEAlgorithm` which means implicit, otherwise mass matrix